### PR TITLE
Reworked workspace files cleaning

### PR DESCRIPTION
**Pull request by sleshchenko**
### What does this PR do?
As spi contract says nothing about creating additional files for workspace, it should not define any mechanism to clean up them. So this pull request makes workspace files cleaning as infrastructure specific feature.
Reworks docker workspace file cleaner according to the new approach.
Adds removing of OpenShift project when workspace is removed.

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/6081

### CODE REVIEW

File: infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/local/projects/RemoveLocalProjectsFolderOnWorkspaceRemove.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  @Inject
+  public RemoveLocalProjectsFolderOnWorkspaceRemove(
+      LocalProjectsFolderPathProvider workspaceFolderPathProvider) {
+    this.workspaceFolderPathProvider = workspaceFolderPathProvider;
+    executor =
+        Executors.newCachedThreadPool(
+            new ThreadFactoryBuilder()
+                .setNameFormat("RemoveLocalProjectsFolderOnWorkspaceRemove-%d")
+                .setUncaughtExceptionHandler(LoggingUncaughtExceptionHandler.getInstance())
+                .setDaemon(true)
```
Comment (garagatyi) [inline infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/local/projects/RemoveLocalProjectsFolderOnWorkspaceRemove.java]: In this case, we may have untraced unremoved folders that litter the system. Should we use non-daemon threads instead?

File: infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/local/projects/RemoveLocalProjectsFolderOnWorkspaceRemove.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  }
+
+  private void deleteRecursiveAsync(String workspaceId, File toRemove) {
+    executor.execute(
+        ThreadLocalPropagateContext.wrap(
+            () -> {
+              boolean success = deleteRecursive(toRemove);
+              if (!success) {
+                LOG.error(
+                    "Failed to remove projects files for workspace with id: '{}'.", workspaceId);
```
Comment (garagatyi) [inline infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/local/projects/RemoveLocalProjectsFolderOnWorkspaceRemove.java]: Please add path to not removed folder too

File: infrastructures/docker/src/test/java/org/eclipse/che/workspace/infrastructure/docker/local/projects/LocalProjectsFolderPathProviderTest.java [hunk 3]
```diff
@@ -8,18 +8,19 @@
...
-package org.eclipse.che.workspace.infrastructure.docker;
+package org.eclipse.che.workspace.infrastructure.docker.local.projects;
 
 import org.mockito.testng.MockitoTestNGListener;
 import org.testng.annotations.Listeners;
 
 @Listeners(value = {MockitoTestNGListener.class})
-public class LocalWorkspaceFolderPathProviderTest {
+public class LocalProjectsFolderPathProviderTest {
+  //TODO Fix tests
```
Comment (garagatyi) [inline infrastructures/docker/src/test/java/org/eclipse/che/workspace/infrastructure/docker/local/projects/LocalProjectsFolderPathProviderTest.java]: Are you going to fix tests in this PR or another one?

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/RemoveProjectOnWorkspaceRemove.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+  private final OpenShiftClientFactory clientFactory;
+
+  @Inject
+  public RemoveProjectOnWorkspaceRemove(OpenShiftClientFactory clientFactory) {
+    this.clientFactory = clientFactory;
+  }
+
+  @Inject
+  public void subscribe(EventService eventService) {
```
Comment (garagatyi) [inline infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/RemoveProjectOnWorkspaceRemove.java:38]: We usually do this with post construct annotation. Can you elaborate on how this solution differs?

File: infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/local/projects/RemoveLocalProjectsFolderOnWorkspaceRemove.java (hunk 1 above)
Comment (sleshchenko) [reply infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/local/projects/RemoveLocalProjectsFolderOnWorkspaceRemove.java]: It was implemented in this way. So I think It makes sense to change it in way you proposed

File: infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/local/projects/RemoveLocalProjectsFolderOnWorkspaceRemove.java (hunk 2 above)
Comment (sleshchenko) [reply infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/local/projects/RemoveLocalProjectsFolderOnWorkspaceRemove.java]: Makes sense. Thanks for remark

File: infrastructures/docker/src/test/java/org/eclipse/che/workspace/infrastructure/docker/local/projects/LocalProjectsFolderPathProviderTest.java (hunk 3 above)
Comment (sleshchenko) [reply infrastructures/docker/src/test/java/org/eclipse/che/workspace/infrastructure/docker/local/projects/LocalProjectsFolderPathProviderTest.java]: Yeah. See PR description

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/RemoveProjectOnWorkspaceRemove.java (hunk 4 above)
Comment (sleshchenko) [reply infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/RemoveProjectOnWorkspaceRemove.java:38]: `@PostConstuct` annotation is functionality that is implemented by ourselves https://github.com/eclipse/che/blob/6e9825c699d34eef6b8bf36fc0194ff871cbee84/core/commons/che-core-commons-inject/src/main/java/org/eclipse/che/inject/CheBootstrap.java#L135

Using of `@Inject` annotation on methods is functionality that is provided by Guice.

As we don't need EventService outside of this method I prefer to inject EventService instance into the method.

File: infrastructures/docker/src/test/java/org/eclipse/che/workspace/infrastructure/docker/local/projects/LocalProjectsFolderPathProviderTest.java (hunk 3 above)
Comment (garagatyi) [reply infrastructures/docker/src/test/java/org/eclipse/che/workspace/infrastructure/docker/local/projects/LocalProjectsFolderPathProviderTest.java]: Sorry, I missed WIP mark.

File: plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftWorkspaceFilesCleaner.java [hunk 5]
```diff
@@ -39,47 +41,61 @@
...
-          }
-        });
+
+    this.deleteQueue = ConcurrentHashMap.newKeySet();
+    this.executor =
+        Executors.newCachedThreadPool(
+            new ThreadFactoryBuilder()
+                .setNameFormat("OpenShiftWorkspaceFilesCleaner-%d")
+                .setUncaughtExceptionHandler(LoggingUncaughtExceptionHandler.getInstance())
+                .setDaemon(true)
```
Comment (garagatyi) [inline plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftWorkspaceFilesCleaner.java]: Make non-daemon, please.

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/RemoveProjectOnWorkspaceRemove.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+import org.eclipse.che.api.workspace.server.event.WorkspaceRemovedEvent;
+import org.eclipse.che.workspace.infrastructure.openshift.OpenShiftClientFactory;
+
+/**
+ * Listener for removing OpenShift project on {@code WorkspaceRemovedEvent}.
+ *
+ * @author Sergii Leshchenko
+ */
+@Singleton
+public class RemoveProjectOnWorkspaceRemove implements EventSubscriber<WorkspaceRemovedEvent> {
```
Comment (garagatyi) [inline infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/RemoveProjectOnWorkspaceRemove.java:28]: Add tests, please

**Comment by garagatyi** [comment]
Looks good! Waiting for tests to approve.

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/RemoveProjectOnWorkspaceRemove.java (hunk 6 above)
Comment (sleshchenko) [reply infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/RemoveProjectOnWorkspaceRemove.java:28]: Done

**Comment by sleshchenko** [comment]
@garagatyi Please take a look my PR again. I've fixed all remarks

**Comment by garagatyi** [comment]
@sleshchenko is it still WIP?

**Comment by sleshchenko** [comment]
@garagatyi No. Removed WIP from the title.

**Review by akorneta — APPROVED** [review]

**Review by garagatyi — APPROVED** [review]
Looks good!

**Pull request merged**